pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,805 @@
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		klog.Warningf("the packet file %s already exists. This may be caused by an unexpected termination, will delete it", filePath)
+		if err := defaultFS.Remove(filePath); err != nil {
+			return nil, nil, err
+		}
+	}
+	file, err := defaultFS.Create(filePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target
+// Pod does not exist on the current Node, the agent on this Node will not perform the capture.
+// In the PacketCapture spec, at least one of `.Spec.Source.Pod` or `.Spec.Destination.Pod`
+// should be set.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) error {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+	var err error
+	pcState := &packetCaptureState{name: name}
+	cleanup := func() {
+		if err != nil {
+			klog.ErrorS(err, "PackageCapture failed", "name", name)
+			err := c.updatePacketCaptureStatus(name, 0, "", err)
+			if err != nil {
+				klog.ErrorS(err, "Failed to update PacketCapture status")
+			}
+		}
+	}
+	defer cleanup()
+	pc, err := c.packetCaptureLister.Get(name)
+	if err != nil {
+		return err
+	}
+	srcIP, dstIp, err := c.parseIPs(pc)
+	if err != nil {
+		return err
+	}
+	// checked before, ensured it's not nil here.
+	device := c.getTargetCaptureDevice(pc)
+	klog.V(2).InfoS("Prepare capture on the current Node", "name", pc.Name, "device", *device)
+	pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	timeout := defaultTimeoutDuration
+	if pc.Spec.Timeout != nil {
+		timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+	}
+	go func() {
+		captureErr := c.performCapture(pc, pcState, *device, srcIP, dstIp, timeout)
+		func() {
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			c.numRunningCaptures -= 1
+			c.captures[name] = packetCapturePhaseCompleted
+		}()
+		updateErr := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", captureErr)
```

Comment:
We should separate the capture goroutine and the worker goroutine's responsibilites clearly and avoid introducing race condition between them.

One way is to use the `captures` map to maintain the capture runtime state of the capture, including its phase and numberCaptured. The capture goroutins updates the state and enqueue the capture to schedule worker to reflect status in API. The worker goroutine is the only one that could update status in API.

Reply:
updated.